### ISSUE TITLE
Import declarations may only appear at top level of a module (possible fix for GNOME 45)

### ISSUE THREAD

**Issue by anmolmathias** [rvaiya/keyd#603]
I'm on Fedora 39 with GNOME 45. I've just reinstalled keyd from source and it works fine with no issues but I can't seem to get application specific remapping work. Running `keyd-application-mapper` outputs the following and hangs:
```
/usr/bin/keyd-application-mapper:469: SyntaxWarning: invalid escape sequence '\W'
  return re.sub('[\W_]+', '-', s).strip('-').lower()
Gnome detected
```
Force quitting it then outputs this:
```
^CTraceback (most recent call last):
  File "/usr/bin/keyd-application-mapper", line 499, in <module>
    mon.run()
  File "/usr/bin/keyd-application-mapper", line 391, in run
    fh = open(self.fifo_path)
         ^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
I can't run the extension itself because of the following error: `import declarations may only appear at top level of a module`

![Screenshot from 2023-10-28 22-23-55](https://github.com/rvaiya/keyd/assets/15323519/53565ed2-1299-4f15-b7e0-dbdf1f444986)

I've tried editing the extension metadata file for compatibility with 45 but it doesn't seem to make a difference. Any ideas what I'm missing? Is there a way to work this without the extension on GNOME somehow?

**Comment by anmolmathias** [comment]
After reading around a bit I think the issue is that with GNOME 45, the syntax to import modules have changed to standard JavaScript modules (ESM). Due to this change and among others like mandatory export of a default class, extension compatibility has been broken with previous shell versions. It's possible the extension will need a rewrite for 45.

**Comment by anmolmathias** [comment]
Update: I tried to refactor `extension.js` to run on GNOME 45 and it seems to work again. I'm posting the refactored code  below for anyone that might want to get the extension working on 45 while a proper rewrite is worked on (albeit @rvaiya  agrees to it). Beware that I've got no understanding of ECMAScript 6 or how keyd-application-mapper actually works; this is just a moving syntax around by reading other code to try to get the gnome-shell to execute the extension relatively error-free (syntactically at least). But it works for now, so whatever.

```

        import Shell from 'gi://Shell';
        import GLib from 'gi://GLib';
        import * as Main from 'resource:///org/gnome/shell/ui/main.js';
        import { Extension, gettext as _ } from 'resource:///org/gnome/shell/extensions/extension.js';

        // We have to keep an explicit reference around to prevent garbage collection :/.
        let file = imports.gi.Gio.File.new_for_path('/home/nml/.local/share/gnome-shell/extensions/keyd/keyd.fifo');
        let pipe = file.append_to_async(0, 0, null, on_pipe_open);

        function send(msg) {
            if (!pipe)
                return;

            try {
                pipe.write(msg, null);
            } catch {
                log('pipe closed, reopening...');
                pipe = null;
                file.append_to_async(0, 0, null, on_pipe_open);
            }
        }

        function on_pipe_open(file, res) {
            log('pipe opened');
            pipe = file.append_to_finish(res);
        }

        export default class KeydExtension extends Extension {
            enable() {
                Shell.WindowTracker.get_default().connect('notify::focus-app', () => {
                    const win = global.display.focus_window;
                    const cls = win ? win.get_wm_class() : 'root';
                    const title = win ? win.get_title() : '';

                    send(`${cls}\t${title}\n`);
                });

                Main.layoutManager.connectObject(
                    'system-modal-opened', () => {
                        send(`system-modal\t${global.stage.get_title()}\n`);
                    },
                    this
                );
                GLib.spawn_command_line_async('keyd-application-mapper -d');
            }

            disable() {
                GLib.spawn_command_line_async('pkill -f keyd-application-mapper');
            }
        }
        
```
Make sure to edit 'metadata.json' to reflect `shell-version` to just `45`:

```
        {
                "name": "keyd",
                "description": "Used by keyd to obtain active window information.",
                "uuid": "keyd",
                "shell-version": [ "45" ]
        }

```

Awaiting feedback.

**Comment by canadaduane** [comment]
Here's a draft PR to bring the helpful additions @anmolmathias has made back to the main branch: https://github.com/rvaiya/keyd/pull/639

Any help in figuring out the error would be appreciated, then we can get this merged :)

**Comment by rvaiya** [comment]
Sorry for taking so long to address this. I believe this should be fixed in 88b9b81b8, though confirmation/further testing would be appreciated since I don't daily drive Gnome.

>  this is just a moving syntax around by reading other code to try to get gnome-shell to execute the extension relatively error-free (syntactically at least). But it works for now, so whatever.

Gnome extensions are notoriously fickle and poorly documented. I had to read a fair bit of obscure code to cobble the extension together in the first place :P.
